cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java

```diff
@@ -0,0 +1,219 @@
+package co.cask.cdap.internal.app.runtime.batch;
+
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import co.cask.cdap.api.Resources;
+import co.cask.cdap.api.data.DatasetInstantiationException;
+import co.cask.cdap.api.data.batch.Split;
+import co.cask.cdap.api.data.stream.StreamBatchReadable;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.mapreduce.MapReduceContext;
+import co.cask.cdap.api.mapreduce.MapReduceSpecification;
+import co.cask.cdap.api.mapreduce.MapReduceTaskContext;
+import co.cask.cdap.api.templates.AdapterSpecification;
+import co.cask.cdap.api.templates.plugins.PluginProperties;
+import co.cask.cdap.api.workflow.WorkflowToken;
+import co.cask.cdap.internal.app.runtime.batch.dataset.MultipleOutputs;
+import org.apache.hadoop.mapreduce.TaskInputOutputContext;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Mapreduce task runtime context which delegates to BasicMapReduceContext for non task-specific methods.
+ * It currently also extends MapReduceContext to support backwards compatability. Mapper and Reducer tasks could
+ * implement ProgramLifeCycle<MapReduceContext> in order to to get a MapReduceContext object.
+ *
+ * @param <KEYOUT>   output key type
+ * @param <VALUEOUT> output value type
+ */
+public class BasicMapReduceTaskContext<KEYOUT, VALUEOUT> implements MapReduceTaskContext<KEYOUT, VALUEOUT>,
+  MapReduceContext {
+
+  private final BasicMapReduceContext basicMapReduceContext;
+  private TaskInputOutputContext<?, ?, KEYOUT, VALUEOUT> hadoopContext;
+  private final MultipleOutputs multipleOutputs;
+
+  public BasicMapReduceTaskContext(BasicMapReduceContext basicMapReduceContext,
+                                   TaskInputOutputContext<?, ?, KEYOUT, VALUEOUT> hadoopContext) {
+    this.basicMapReduceContext = basicMapReduceContext;
+    this.hadoopContext = hadoopContext;
+    this.multipleOutputs = new MultipleOutputs(hadoopContext);
+  }
+
+  @Override
+  public <K, V> void write(String namedOutput, K key, V value) throws IOException, InterruptedException {
+    multipleOutputs.write(namedOutput, key, value);
+  }
+
+
+  @Override
+  public void write(KEYOUT key, VALUEOUT value) throws IOException, InterruptedException {
+    hadoopContext.write(key, value);
+  }
+
+  public void close() throws IOException, InterruptedException {
```

Comment:
I thought you said not to have it.
https://github.com/caskdata/cdap/pull/3740#discussion_r38240675